include/seqan3/range/view/kmer_hash.hpp

```diff
@@ -12,47 +12,645 @@
+
+    /*!\brief Iterator for calculating hash values via a given seqan3::shape.
+     * \tparam urng_t Type of the text. Must model std::ForwardRange. Reference type must model seqan3::Semialphabet.
+     *
+     * \details
+     *
+     * The shape_iterator can be used to iterate over the hash values of a text. A shape_iterator needs an iterator of
+     * the text and a seqan3::shape that defines how to hash the text.
+     *
+     * Depending on the type of the iterator passed to the shape_iterator, different functionality is available:
+     *
+     * | Concept modelled by passed text iterator                     | Available functions             |
+     * |--------------------------------------------------------------|---------------------------------|
+     * | \ref std::ForwardIterator "std::ForwardIterator"             | \ref shape_iterator_comparison "Comparison operators"<br>\ref operator++ "Pre-increment (++it)"<br>\ref operator++(int) "Post-increment (it++)"<br>\ref operator* "Indirection operator (*it)" |
+     * | \ref std::BidirectionalIterator "std::BidirectionalIterator" | \ref operator-- "Pre-decrement (--it)"<br>\ref operator--(int) "Post-decrement (it--)" |
+     * | \ref std::RandomAccessIterator "std::RandomAccessIterator"   | \ref operator+= "Forward (it +=)"<br>\ref operator+ "Forward copy (it +)"<br>\ref operator-= "Decrement(it -=)"<br>\ref shape_iterator_operator-decrement "Decrement copy (it -)"<br>\ref shape_iterator_operator-difference "Difference (it1 - it2)"<br>\ref operator[] "Subscript (it[])" |
+     *
+     * When using a gapped seqan3::shape, the `0`s of the seqan3::shape are virtually removed from the hashed k-mer.
+     * Note that any shape is expected to start with a `1` and end with a `1`.
+     *
+     * ### Implementation detail
+     *
+     * To avoid dereferencing the sentinel when iterating, the shape_iterator computes the hash value up until
+     * the second to last position and performs the addition of the last position upon
+     * access (\ref operator* and \ref operator[]).
+     */
+    template <typename rng_t>
+    class shape_iterator
+    {
+    private:
+        //!\brief The iterator type of the underlying range.
+        using it_t = std::ranges::iterator_t<rng_t>;
+        //!\brief The sentinel type of the underlying range.
+        using sentinel_t = std::ranges::sentinel_t<rng_t>;
+
+    public:
+        /*!\name Associated types
+         * \{
+         */
+        //!\brief Type for distances between iterators.
+        using difference_type = typename it_t::difference_type;
+        //!\brief Value type of this iterator.
+        using value_type = size_t;
+        //!\brief The pointer type.
+        using pointer = void;
+        //!\brief Reference to `value_type`.
+        using reference = value_type;
+        //!\brief Tag this class depending on which concept it_t models.
+        using iterator_category = std::conditional_t<std::ContiguousIterator<it_t>,
```

Comment:
```suggestion
        using iterator_category = std::input_iterator_tag;
        using iterator_concept = std::conditional_t<std::ContiguousIterator<it_t>,
```

Because it returns values. We need to fix this for a few views.